### PR TITLE
Add active surveys to the warning message when deleting a channel

### PR DESCRIPTION
**Pull request by ismaelbej**
### Changes

* Add query to retrieve running surveys for a provider and baseUrl
* Change modal to show active surveys
* Add unit test for the endpoints

#### Searching for active surveys

Screenshot 2024-11-19 at 6 39 55 PM

#### No active surveys

Screenshot 2024-11-19 at 6 39 02 PM

#### List of active surveys

Screenshot 2024-11-21 at 7 48 45 PM

### CODE REVIEW

File: assets/js/components/channels/ChannelIndex.jsx [hunk 1]
```diff
@@ -37,6 +57,31 @@ class ChannelIndex extends Component<any> {
...
+          modalIndex: index,
+        })
+      })
+      .catch(() => {
+        this.setState({
+          modalLoading: false,
+          modalSurveys: [],
+          modalProvider: provider,
+          modalIndex: index,
+        })
```
Comment (matiasgarciaisaia) [inline assets/js/components/channels/ChannelIndex.jsx:83]: This is silently dismissing an exception. Shouldn't we let it throw (so the user gets an automatic error message popup)? Or set `modalSurveys` to `null` ("error, beware") and tell it apart from an empty list ("no active channels, safe to delete provider").

File: assets/js/components/channels/ProviderModal.jsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      showCancel
+    >
+      <div>
+        <p>{t("Do you want to delete the channels provided by {{name}}?", { name })}</p>
+
+        <div className="provider-surveys">
+          {loading ? <span>{t("Searching active surveys...")}</span> :
+            surveys.length == 0 ? <span>{t("No active surveys")}</span> :
+              <div>
+                <span>{t("Active surveys")}</span>
```
Comment (matiasgarciaisaia) [inline assets/js/components/channels/ProviderModal.jsx:34]:
```suggestion
                <span>{t("These surveys are active, using channels from this provider. Deleting the channels will interrupt the surveys.")}</span>
```

May be a better wording, but I'd try to explain a bit more why this is relevant to the user.

**Review by matiasgarciaisaia — CHANGES_REQUESTED** [review]
Let's add access control to the endpoint.

File: assets/js/components/channels/ChannelIndex.jsx (hunk 1 above)
Comment (matiasgarciaisaia) [reply assets/js/components/channels/ChannelIndex.jsx]: > I'm showing a generic error message an logging the error to the browser's console

Isn't that Surveda's default behaviour if we simply don't catch the exception? There's some standard component there that shows exceptions as a notification bubble, not sure if that applies here too.

Comment (ismaelbej) [reply assets/js/components/channels/ChannelIndex.jsx]: I've removed the catch handler and changes related.

File: lib/ask/survey.ex [hunk 3]
```diff
@@ -530,10 +531,16 @@ defmodule Ask.Survey do
...
+  def with_active_channels(user_id, provider, base_url) do
     query =
       from s in Survey,
         where: s.state == :running,
+        join: pm in ProjectMembership,
+        on: pm.project_id == s.project_id and pm.user_id == ^user_id,
+        select: s
+
+    query =
+      from s in subquery(query),
```
Comment (matiasgarciaisaia) [inline lib/ask/survey.ex:543]:
```suggestion
```

Doesn't it work all together in a single query?

**Review by matiasgarciaisaia — APPROVED** [review]
It's usually a good idea to avoid DB subqueries if possible.

If the subquery is actually needed, let's merge as is!

**Pull request merged**